### PR TITLE
HHH-10289 - CPU performance regression in StatefulPersistenceContext.…

### PR DESCRIPTION
**Pull request by johnaoahra80**
…addEntry()

### CODE REVIEW

**Comment by Sanne** [comment]
Hi @johnaoahra80 
these changes are extremely puzzling; is it an experiment or did you verify their effect already?

Would it be possible to add many more comments to explain exactly why some choices were made?

I see there is a general warning "don't ever touch this again", but we'll probably need to make options at some point and it would be nice to know how to maintain it.

**Comment by johnaoahra80** [comment]
@Sanne I will provide comments on the PR.  We identified the issues using cpu profiling using perf, and have verified the changes have made an impact on cpu using perf and running against our benchmark.  The main change relates to polymorphic calls to interface methods, and what we have done to is cast the interface to the correct implementation where we know which implementation to use, the jvm does not need to resolve which implementation of the interface to use, which requires more cpu cycles.  The other changes reduce the size of methods to increase the number of methods inlined.

File: hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java [hunk 1]
```diff
@@ -489,13 +489,29 @@ public EntityEntry addEntry(
...
+					disableVersionIncrement,
+					lazyPropertiesAreUnfetched,
+					this
+			);
 		}
 		else {
-			final EntityEntryFactory entityEntryFactory = persister.getEntityEntryFactory();
-			e = entityEntryFactory.createEntityEntry(
+			e = ((ImmutableEntityEntryFactory) persister.getEntityEntryFactory()).createEntityEntry(
 					status,
```
Comment (johnaoahra80) [inline hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java]: We are casting the interface to the known implementation to prevent the JVM from having to resolve which implementations method .createEntityEntry() to call.  This removes the need make calls via polymorphic call sites and reverts the perf regression introduced by having more than one implementation of the EntityEntryFactory interface.

File: hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java [hunk 2]
```diff
@@ -511,12 +527,22 @@ public EntityEntry addEntry(
...
+			final Object entity,
+			final Status status) {
+
+		((ManagedEntity)entity).$$_hibernate_getEntityEntry().setStatus( status );
+		entityEntryContext.addEntityEntry( entity, ((ManagedEntity)entity).$$_hibernate_getEntityEntry() );
+
+		setHasNonReadOnlyEnties( status );
+		return ((ManagedEntity)entity).$$_hibernate_getEntityEntry();
+	}
+
```
Comment (johnaoahra80) [inline hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java]: Refactored method to promote method inling

File: hibernate-core/src/main/java/org/hibernate/event/internal/DefaultLoadEventListener.java [hunk 3]
```diff
@@ -612,10 +629,78 @@ protected Object loadFromSecondLevelCache(
...
+		else {
+			makeEntityCircularReferenceSafe(referenceCacheEntry, entityId, session, entity, entityKey);
+			//PostLoad is needed for EJB3
+			EventListenerGroup<PostLoadEventListener> evenListenerGroup = getEvenListenerGroup(session);
+
+			if(!evenListenerGroup.isEmpty()) {
+				postLoad(session, evenListenerGroup.listeners(), entity, entityId, persister);
+			}
+			return entity;
+		}
```
Comment (johnaoahra80) [inline hibernate-core/src/main/java/org/hibernate/event/internal/DefaultLoadEventListener.java]: Not allocating an eventListernerGroup iterator if the listeners collection is empty

**Comment by johnaoahra80** [comment]
Comparing before and after this PR is applied we see;

For one metric in our benchmark (the metric that is mostly effected by these changes);
Avg. response latency;
Before: 4.894s
After: 0.023s

Overall CPU utilisation;

Before: 95.25%
After: 59.83%

Cpu utilisation for 
org/hibernate/event/internal/DefaultLoadEventListener:.loadFromSecondLevelCache()

Before: 28.74% Cpu
After: 7.31%

File: hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java (hunk 1 above)
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java]: Understood, but I think a comment about this should be added to the code to help who's going to read (and potentially modifyi) in the future.

Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java]: so doing an explicit branch on an "instanceof" evaluation, followed by casting, is more efficient than having the JVM do an implicit "instanceof" evaluation ? Looks like hotspot needs to do the same things - for example to verify the cast is valid - so I'm not understanding why our explicit code is more efficient.

Comment (johnaoahra80) [inline hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java]: The JVM does need to type check, but before it can type check and cast, it needs to resolve the correct instance using a vtable lookup. The vtable lookup takes 3 bytecode instructions to complete vs 1 bytecode instruction for the direct method call (see https://wiki.openjdk.java.net/display/HotSpot/VirtualCalls).  By casting to a concrete implementation, the JVM does not need to interpret which concrete method to call with each virtual call. We are hard coding that lookup.

This also allows the JVM to perform further optimisations, e.g. inlining method calls.

We can measure the cpu impact using perf, lookig at the method call org/hibernate/event/internal/DefaultLoadEventListener:.loadFromSecondLevelCache()

Before: 28.74% CPU
After: 7.31% CPU

Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java]: clearly your results are amazing, so I'll have to believe it :)

Still I'm confused. Let me rephrase that and see if I understand:
with your patch, the type check only happens once at casting and takes 3 bytecodes there; then costs only 1 for each method call. 
W/o your patch, there's no upfront vtable lookup but there's one for each method invocation.
Don't you think that hotspot should be able to rewrite the generated bytecode to keep the first vtable resolution around?

Could you add some notes in the code, or there's risk for these optimisations to be removed in near future.

**Comment by Sanne** [comment]
Those figures are way too good to be belieavable for such a small patch, in such a large picture :-D

**Comment by johnaoahra80** [comment]
@Sanne I think it is worth noting that these are improvements in comparison to the regression that was introduced by introducing a new implementation of EntityEntryFactory. We are still using slightly more CPU cycles compared to running ORM4.2 in EAP6.4;

I have chosen a slightly different method call to compare as more methods are inlined in ORM4.2.
org/hibernate/event/internal/DefaultLoadEventListener:.onLoad()

EAP6.4 (ORM4.2.16.Final) ; 7.18% CPU
WF10 (ORM5.0.1.Final); 29.52% CPU
WF10 (ORM5.0.1.Final + patch); 7.76% CPU

**Comment by stalep** [comment]
@Sanne Hi, early last week we were trying to understand the drop in performance when we used the re-using entity-entry patch that @johnaoahra80 wrote a long time ago. At the fairly low IR we ran at the time the difference was huge. The only explanation we could find was that we had created a bimorphic method call on what used to be a monomorphic method call. When we changed the method call back to being monomorphic we saw a huge gain in performance again.
We have numerous benchmark runs, profiles showing these findings.

**Comment by Sanne** [comment]
I understand the patch is great, I didn't mean to suggest it shouldn't be merged. Just trying to understand as I'm surprised by the huge impact such a small thing can have.
If to merge or not, that's really up to Steve only not me. +1 from me if that helps ;)

I still think some more comments should be added to the code, but if you're in a rush I guess that could be done later.

File: hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java (hunk 1 above)
Comment (johnaoahra80) [inline hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java]: With the patch, there is a type check and a non-local jump
Without the patch, this is a type check, 3 memory references and 2 non local jumps
The JVM can not cache the vtable resolution as there is more than one class loaded that implements the virtual method, and needs to resolve this everytime.  If it were to cache the resolution, you would have method calls being dispatched to the wrong class.

When there is only 1 implementation of the virtual method, the JVM knows that there can only be one concerete method, so replaces the virtual method call with a direct call and removes the vtable lookup.  This was the case in ORM4.2.  When we introduced another implementation of the EntityEntryFactory interface, the JVM can no longer make this optimisation.

Another effect is, in 4.2, the JVM could substitute the virtual method call for a direct method call, then it was able to inline that method at call sites.  With 2 concrete methods, the JVM is unable to inline the method calls at the call sites as it needs to resolve which implemenation to use

**Comment by sebersole** [comment]
Applied upstream.  I did reformat the in-line comments some along with some minor spelling and grammar fixes.
